juju/juju
# Caas storage detach

## Description of change

One liner - support detaching k8s storage instead of forcing it to be destroyed when removing a unit/application.

There's a large amount of iaas storage infrastructure for iaas clouds which was modified to also run with caas clouds. The essential stack is the same, but there's a number of differences to cater for:
- caas storage created/attached when the pod is provisioned, not separately
- the workers to monitor model scoped and unit scoped storage all run in the model manifold, not split across each application agent
- the workers just handle storage cleanup when things are deleted in the juju model

The work is broken up into several commits:

1. Allow caas storage to be detached instead of forcing destruction
Just a cleanup to revert the hard coding to destroy caas storage

2. Tweak state storage watchers and struct attachment tag to allow for watching unit storage
Just modify the regexp used with the existing state watchers to cater for unit attached storage as well as model attached storage.

3. Adapt the StorageProvisioner API facade
Modify the facade to support caas storage.

4. Adapt the StorageProvisioner worker
Modify the facade to support caas storage.

5. Add filesystem storage source to k8s broker
Implement some k8s FilesystemSource methods to delete volumes

6. Fix use of test kubernetes charms
We are now a lot stricter about loading fake k8s charms from the test repo.

## QA steps

deploy a k8s charm with storage
remove a unit and ensure the volume/filesystem are shown as detached
remove a unit is --destroy-storage ensure the volume/filesystem are deleted
